### ISSUE TITLE
addons and helpers in 0.2.2

### ISSUE THREAD

**Issue by dagda1** [ember-cli/ember-cli#3796]
I think this is a bug in 0.2.2:

What is the correct way of placing your helper files and also where should they go with respect to the resolver finding them from an addon ember-cli project?

I am running ember-cli 0.2.2.

I generated an helper from an addon project with:
```
ember g helper display-helper
```
The generator placed the file in ```app/helpers``` which seemed wrong to me, I would have thought that it should have been placed in addon helpers. I moved the file to ```addon/helpers```.

I tried adding a file in ```app/helpers/display-helper.js``` that imports and exports the helper from the file in the addons directory but from what I can tell, no files are being automatically loaded from the ```app/helpers``` directory if it is an addon project.

When I ran ```ember test``` I get the following output:
>✘ Error: Assertion Failed: A helper named 'displayHelper' could not be found

The only way I get this helper to be found by the resolver is to add an import that references the helper in a component file that is in the addon directory that is using the helper like this:
```
import displayHelper from '../helpers/display-helper';
```
This does not seem correct, I would have thought the resolver would have found this automatically?

Also even if I have the reference, the following generated code ends up with the same error message as above:
```
import Ember from 'ember';

var get = Ember.get;

function displayHelper(context, searchPath) {
  return new Ember.Handlebars.SafeString(get(context, searchPath));
}

export default Ember.Handlebars.makeBoundHelper(displayHelper);
```

I had to change the code to this:
```
import Ember from 'ember';

var get = Ember.get;

export default Ember.Handlebars.registerBoundHelper('displayHelper', function displayHelper(context, searchPath) {
  return new Ember.Handlebars.SafeString(get(context, searchPath));
});
```
So to sum up, I have to have this line in a component whose template uses the helper:
```
import displayHelper from '../helpers/display-helper';
```
And I have to use ```registerBoundHelper``` and not ```makeBoundHelper``` like the docs say or the helper cannot be found.

**Comment by rwjblue** [comment]
The invalid generation was just fixed in https://github.com/ember-cli/ember-cli/pull/3690.


You can see the expected generation in [this test](https://github.com/ember-cli/ember-cli/blob/2e64faa354f9d4f2a1c0202f91bd5f325345af2e/tests/acceptance/addon-generate-test.js#L177-L198).

**Comment by rwjblue** [comment]
Your scenario should be:

```javascript
// addon/helpers/display-helper.js
import Ember from 'ember';

var get = Ember.get;

function displayHelper(context, searchPath) {
  return new Ember.Handlebars.SafeString(get(context, searchPath));
}

export default Ember.Handlebars.makeBoundHelper(displayHelper);
```

```javascript
// app/helpers/display-helper.js
import helper from 'my-addon-name/helpers/display-helper';

export default helper;
```

**Comment by dagda1** [comment]
I don't think this is incorrect on both counts, the generations appear to be broke and one other person has confirmed as I reported in #3744.  Nobody has replied to this.  Although the tests are going into the right places, I don't think the resolver is looking for them in the ```addon``` directory.

Also, the files in ```app/helpers``` do not appear to be loaded, if I create a file in ```app/helpers``` and add some non valid javascript to it like this:
```
fdsjfsldjfsods;

```
I can still run my tests and they all pass which leads me to believe that the file is never, ever loaded.

I think there are quite a few problems with the 0.2.2 release related to the #3690.

**Comment by rwjblue** [comment]
The changes from https://github.com/ember-cli/ember-cli/pull/3690 had just landed when I commented last, did you test them?  If you did, I'd love to see a demo repo that shows the issues you see still existing.

> Although the generated files are going into the right places, I don't think the resolver is looking for them in the addon directory.

The resolver should not be looking in the `addon` directory, that is why you have to import and re-export in the `app` directory (so that the resolver will find the files).

> Also, the files in app/helpers do not appear to be loaded

Nothing is loaded unless it is looked for.  Adding a file at `app/helpers/foo-bar.js` will never be loaded unless you use that helper in a template (e.g. `{{foo-bar stuff}}`).

**Comment by dagda1** [comment]
That file does not get loaded, I am using the helper in a template that is being run from a test in a [test](https://github.com/dagda1/emberx-autosuggest/blob/master/addon/templates/components/x-autosuggest.hbs#L6) but if I change the code the way you said then I get an error saying that the helper cannot be found and if I add the file in ```app/helpers``` it does not get loaded.

The only way I can get the tests to pass is to add the import statement in the component [here](https://github.com/dagda1/emberx-autosuggest/blob/master/addon/components/x-autosuggest.js#L3) which is obviously wrong.

Are you saying the fix from last night will sort these issues?

If so, how can I overrite my ember-cli considering that I am running 0.2.2 and the version on github is 0.2.2?

**Comment by dagda1** [comment]
@rwjblue unless I am mistaken which is very possible, this is still a bug.  I have updated [this](https://github.com/dagda1/ember-flow) to ember 0.2.3 and the bug still remains.

I have this [file](https://github.com/dagda1/emberx-autosuggest/blob/master/app/helpers/display-helper.js) in ```app/helpers``` and I have this [file](https://github.com/dagda1/emberx-autosuggest/blob/master/addon/helpers/display-helper.js) in ```addon/helpers```.

The helper is referenced in this [template](https://github.com/dagda1/emberx-autosuggest/blob/master/addon/templates/components/x-autosuggest.hbs#L6).

But when I run the tests, I get this output:

> Error: Assertion Failed: A helper named 'displayHelper' could not be found

I think the bug still remains.

**Comment by rwjblue** [comment]
The resolver requires helpers to be dasherized (which yours is), but you are using it in camel case.  Change your template to use `display-helper`, and it should work properly.

**Comment by dagda1** [comment]
@rwjblue works a treat, thanks for your patience

**Comment by rwjblue** [comment]
Awesome, glad we got it straightened out.
